File: 279-perfect-squares/perfect-squares.cpp
class Solution {
public:
    int numSquares(int n) {
        vector<int> val;
        for(int i=1;i*i<=n;++i){
            val.push_back(i*i);
        }
        vector<int> dp(n+1,1e9);
        dp[0]=0;
        for(int i=0;i<n;++i){
            if(dp[i]==1e9) continue;
            for(int v: val){
                if(i>n-v) break;
                dp[i+v]=min(dp[i+v],1+dp[i]);
            }
        }
        return dp[n];
    }
};

File: 279-perfect-squares/perfect-squares.java
class Solution {
    private int numSquaresUtil(int n, int[] dp){
        if(n==0 || n==1) return n;
        if(dp[n]==-1){
            int smallest= n;
            for(int i=1;i*i<=n;i++){
                smallest= Math.min(smallest, numSquaresUtil(n-i*i, dp));
            }
            dp[n]= 1+smallest;
        }
        return dp[n];
    }
    public int numSquares(int n) {
        int[] dp = new int[n+1];
        Arrays.fill(dp, -1);
        return numSquaresUtil(n, dp);
    }
}